halo2/src/recursion/endoscale.rs

```diff
@@ -0,0 +1,102 @@
+//! Gadget for endoscaling.
+
+use ff::PrimeFieldBits;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::Error,
+};
+use pasta_curves::arithmetic::CurveAffine;
+
+mod chip;
+pub mod primitive;
+
+/// Instructions to map bitstrings to and from endoscalars.
+pub trait EndoscaleInstructions<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits,
+{
+    /// A bitstring to be used in endoscaling.
+    type Bitstring: std::fmt::Debug + Clone;
+    /// The maximum number of bits that can be represented by [`Self::Bitstring`].
+    const MAX_BITSTRING_LENGTH: usize;
+
+    /// Witnesses a slice of bools as a vector of [`Self::Bitstring`]s.
+    fn witness_bitstring(bits: &[bool]) -> Vec<Self::Bitstring>;
+
+    /// Computes commitment (Alg 1) to a variable-length bitstring using the endoscaling
+    /// algorithm. Uses a fixed base.
+    ///
+    /// The bitstring is decomposed into pairs of bits using a running sum outside of
+    /// this gadget.
+    ///
+    /// # Panics
+    /// Panics if `NUM_BITS` is larger than `MAX_BITSTRING_LENGTH`.
+    /// Panics if `NUM_BITS` is an odd number.
+    #[allow(clippy::type_complexity)]
+    fn endoscale_fixed_base<L: Layouter<C::Base>, const NUM_BITS: usize, const NUM_WINDOWS: usize>(
+        &self,
+        layouter: L,
+        base: C,
```

Comment:
Currently, there is no way to constrain in the circuit that multiple invocations of `endoscale_fixed_base` use the same fixed base (it can happen incidentally, but only if the caller passes the same value through to be constant-assigned; a mistake here couldn't be caught in an automated way). Is this intentional (i.e. are the bases supposed to be unconnected)? If not, then a separate associated type, and a corresponding "assign this type from a constant" would be good (and probably what we actually want is just #334).